crates/viewer/re_dataframe_ui/src/filters/string.rs

```diff
@@ -0,0 +1,299 @@
+                )
+                .show_ui(ui, |ui| {
+                    for possible_op in crate::filters::StringOperator::ALL {
+                        if ui
+                            .button(
+                                SyntaxHighlightedBuilder::keyword(&possible_op.to_string())
+                                    .into_widget_text(ui.style()),
+                            )
+                            .clicked()
+                        {
+                            self.operator = *possible_op;
+                        }
+                    }
+                });
+        });
+
+        let response = ui.text_edit_singleline(&mut self.query);
+
+        if popup_just_opened {
+            response.request_focus();
+        }
+
+        action_from_text_edit_response(ui, &response)
+    }
+}
+
+pub fn is_supported_string_datatype(data_type: &DataType) -> bool {
+    matches!(
+        data_type,
+        DataType::Utf8 | DataType::Utf8View | DataType::LargeUtf8
+    )
+}
+
+/// Custom UDF for performing filtering on a column of strings, with support for list columns.
+///
+/// This UDF converts both the haystack and the needle to lowercase before performing the queries.
+#[derive(Debug, Clone)]
+struct StringFilterUdf {
+    needle: String,
+    operator: StringOperator,
+    signature: Signature,
+}
+
+impl StringFilterUdf {
+    fn new(filter: &StringFilter) -> Self {
+        let signature = Signature::one_of(
+            vec![
+                TypeSignature::String(1),
+                TypeSignature::ArraySignature(ArrayFunctionSignature::Array {
+                    arguments: vec![ArrayFunctionArgument::Array],
+                    array_coercion: None,
+                }),
+            ],
+            Volatility::Immutable,
+        );
+
+        Self {
+            needle: filter.query.to_lowercase(),
+            operator: filter.operator,
+            signature,
+        }
+    }
+
+    fn is_valid_input_type(data_type: &DataType) -> bool {
+        match data_type {
+            DataType::List(field) | DataType::ListView(field) => {
+                // Note: we do not support double nested types
+                is_supported_string_datatype(field.data_type())
+            }
+
+            //TODO(ab): support other containers
+            _ => is_supported_string_datatype(data_type),
+        }
+    }
+
+    fn invoke_primitive_array(&self, array: &ArrayRef) -> DataFusionResult<BooleanArray> {
+        if !is_supported_string_datatype(array.data_type()) {
+            return exec_err!("Unsupported data type {}", array.data_type());
+        }
+
+        // We need to convert the haystack to lowercase first. We delegate this task to the existing
+        // datafusion `lower` UDF.
+        //
+        // Note that this _must_ happen here (and, e.g., not at the `Expr` level), because `lower`
+        // does not support nested datatypes such as lists.
```

Comment:
curious, could there be benefits to do it on the expression level iff it is not nested?

Reply:
Much the contrary actually.

Datafusion's `Expr` implementation of `contains` is identical to mine (eg lowers to using arrow's kernel), BUT does not handle the array vs. scalar scenario we have here. So it creates a vector with a duplicated version of the scalar with the same length at the column.

This is obviously bad for performance (not that we wouldn't have other problem in cases where that becomes noticeable).

Relevant slack thread: https://rerunio.slack.com/archives/C04DLS7CCE4/p1758809948817929